### PR TITLE
Fix #65: add grant_type option for Consent

### LINKED ISSUE #65: Add grant_type option to switch consent uri
Depending on the grant type (Authorization Grant vs Implicit Grant), the consent.response_type must be code or token.

To allow this dynamic value, the grant_type must be declared in the configuration

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #65
| License       | MIT

### CODE REVIEW

File: doc/configure-the-bundle.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ docusign:
         private_key: "YourPrivateKey"
         integration_key: "YourIntegrationKey"
         user_guid: "YourUserId"
+        grant_type: authorization_code # Grant type (authorization_code or implicit)
```
Comment (GregoireHebert) [inline doc/configure-the-bundle.md]:
```suggestion
        response_type: "authorization_code" # Grant type (authorization_code or implicit)
```

File: doc/configure-docusign.md [hunk 2]
```diff
@@ -48,6 +48,7 @@ docusign:
         private_key: "Path/To/Your/PrivateKey.pem"
         integration_key: "YourIntegrationKey"
         user_guid: "YourUserId"
+        grant_type: authorization_code # Grant type (authorization_code or implicit)
```
Comment (GregoireHebert) [inline doc/configure-docusign.md]:
```suggestion
        response_type: "authorization_code" # Grant type (authorization_code or implicit)
```

File: src/Controller/Consent.php [hunk 3]
```diff
@@ -22,25 +22,30 @@ class Consent
...
+        $this->grantType = $grantType;
     }
 
     public function __invoke(): RedirectResponse
     {
         return new RedirectResponse(sprintf(
-            '%s?response_type=token&scope=signature%%20impersonation&client_id=%s&redirect_uri=%s',
+            '%s?response_type=%s&scope=signature%%20impersonation&client_id=%s&redirect_uri=%s',
             $this->consentUri,
+            self::GRANT_TYPE_AUTHORIZATION_CODE === $this->grantType ? 'code' : 'token',
```
Comment (GregoireHebert) [inline src/Controller/Consent.php]: Might be better using an array of associated types, it'll be easier to extends later and it avoid a ternary operation. 

```suggestion
           $this->responseType,
```

File: features/Kernel.php [hunk 4]
```diff
@@ -149,6 +149,7 @@ protected function configureContainer(ContainerBuilder $c, LoaderInterface $load
                 'private_key' => '%kernel.project_dir%/var/jwt/docusign.pem',
                 'integration_key' => $_SERVER['DOCUSIGN_INTEGRATION_KEY'],
                 'user_guid' => $_SERVER['DOCUSIGN_USER_GUID'],
+                'grant_type' => 'authorization_code',
```
Comment (GregoireHebert) [inline features/Kernel.php]:
```suggestion
                'response_type' => 'authorization_code',
```

File: src/Controller/Consent.php [hunk 5]
```diff
@@ -22,25 +22,30 @@ class Consent
...
     public const DEMO_CONSENT_URI = 'https://account-d.docusign.com/oauth/auth';
     public const CONSENT_URI = 'https://account.docusign.com/oauth/auth';
+    public const GRANT_TYPE_AUTHORIZATION_CODE = 'authorization_code';
+    public const GRANT_TYPE_IMPLICIT = 'implicit';
 
     private $envelopeBuilder;
     private $router;
     private $consentUri;
     private $integrationKey;
+    private $grantType;
```
Comment (GregoireHebert) [inline src/Controller/Consent.php]:
```suggestion
    private $responseType;
```

File: src/DependencyInjection/DocusignExtension.php [hunk 6]
```diff
@@ -131,6 +131,7 @@ public function load(array $configs, ContainerBuilder $container): void
                     ->setAutowired(true)
                     ->setPublic(true)
                     ->setArguments([
+                        '$grantType' => $value['auth_jwt']['grant_type'],
```
Comment (GregoireHebert) [inline src/DependencyInjection/DocusignExtension.php]:
```suggestion
                        '$responseType' => $value['auth_jwt']['response_type'],
```

File: src/Controller/Consent.php [hunk 7]
```diff
@@ -22,25 +22,30 @@ class Consent
 {
     public const DEMO_CONSENT_URI = 'https://account-d.docusign.com/oauth/auth';
     public const CONSENT_URI = 'https://account.docusign.com/oauth/auth';
+    public const GRANT_TYPE_AUTHORIZATION_CODE = 'authorization_code';
```
Comment (GregoireHebert) [inline src/Controller/Consent.php]:
```suggestion
    public const RESPONSE_TYPES = ['authorization_code' => 'code', 'implicit' => 'token'];
```

File: src/Controller/Consent.php [hunk 8]
```diff
@@ -22,25 +22,30 @@ class Consent
...
+    private $grantType;
 
-    public function __construct(EnvelopeBuilderInterface $envelopeBuilder, RouterInterface $router, string $consentUri, string $integrationKey)
+    public function __construct(EnvelopeBuilderInterface $envelopeBuilder, RouterInterface $router, string $consentUri, string $integrationKey, string $grantType)
     {
         $this->envelopeBuilder = $envelopeBuilder;
         $this->router = $router;
         $this->consentUri = $consentUri;
         $this->integrationKey = $integrationKey;
+        $this->grantType = $grantType;
```
Comment (GregoireHebert) [inline src/Controller/Consent.php]: And in the constructor: 

```suggestion
        if (!isset(self::REPONSE_TYPES[$responseType])) {
            throw new PutYourExceptionHere();
        }
        
        $this->responseType = self::REPONSE_TYPES[$responseType];
```

File: src/DependencyInjection/Configuration.php [hunk 9]
```diff
@@ -114,6 +114,12 @@ public function getConfigTreeBuilder(): TreeBuilder
                                 ->info('Obtain your user UID (also called API username) from DocuSign Admin > Users > User > Actions > Edit')
                             ->end()
                             ->integerNode('ttl')->defaultValue(3600)->info('Token TTL in seconds (default: 3600)')->end()
+                            ->enumNode('grant_type')
```
Comment (GregoireHebert) [inline src/DependencyInjection/Configuration.php]:
```suggestion
                            ->enumNode('response_type')
```

File: doc/configure-the-bundle.md (hunk 1 above)
Comment (vincentchalamon) [reply doc/configure-the-bundle.md]: The response type is not the same thing than grant type.
* Grant type: the grant type selected in DocuSign (`authorization code grant type` vs `implicit grant type`)
* Response type: `code` vs `token` according to the grant type selected

If we set the response type here, the user would have to search in the DocuSign documentation which one to use according to its grant type. Whereas the user already knows which grant type he/she selected.

Comment (GregoireHebert) [reply doc/configure-the-bundle.md]: ok I see, agreed for the configuration key but everywhere else it should be response_type

Comment (vincentchalamon) [reply doc/configure-the-bundle.md]: Fixed

File: src/DependencyInjection/DocusignExtension.php [hunk 10]
```diff
@@ -127,10 +128,15 @@ public function load(array $configs, ContainerBuilder $container): void
                 ]);
 
             if (!empty($value['auth_jwt'])) {
+                if (!isset(Consent::RESPONSE_TYPE[$value['auth_jwt']['grant_type']])) {
+                    throw new InvalidGrantTypeException('Grant type '.$value['auth_jwt']['grant_type'].' is not valid. '.'Please select one of the followings: authorization_code, implicit.');
```
Comment (GregoireHebert) [inline src/DependencyInjection/DocusignExtension.php]: `implode(array_keys(self::RESPONSE_TYPE));` ?

**Comment by GregoireHebert** [comment]
thanks @vincentchalamon :D

**Pull request merged**